src/libraries/System.Private.CoreLib/src/System/Threading/Lock.cs

```diff
@@ -0,0 +1,1194 @@
+                    if (stateBeforeUpdate == state)
+                    {
+                        if (result == TryLockResult.Spin && !state.HasAnySpinners)
+                        {
+                            isFirstSpinner = true;
+                        }
+                        return result;
+                    }
+
+                    state = stateBeforeUpdate;
+                }
+            }
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public static TryLockResult TryLockInsideSpinLoop(Lock lockObj)
+            {
+                // This method is called from inside a spin loop, it must unregister the spinner if the lock is acquired
+
+                var state = new State(lockObj);
+                while (true)
+                {
+                    Debug.Assert(state.HasAnySpinners);
+                    if (!state.ShouldNonWaiterAttemptToAcquireLock)
+                    {
+                        return state.ShouldNotPreemptWaiters ? TryLockResult.Wait : TryLockResult.Spin;
+                    }
+
+                    State newState = state;
+                    newState.SetIsLocked();
+                    newState.DecrementSpinnerCount();
+
+                    State stateBeforeUpdate = CompareExchange(lockObj, newState, state);
+                    if (stateBeforeUpdate == state)
+                    {
+                        return TryLockResult.Locked;
+                    }
+
+                    state = stateBeforeUpdate;
+                }
+            }
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public static TryLockResult TryLockAfterSpinLoop(Lock lockObj)
+            {
+                // This method is called at the end of a spin loop, it must unregister the spinner always and acquire the lock
+                // if it's available. If the lock is available, a spinner must acquire the lock along with unregistering itself,
+                // because a lock releaser does not wake a waiter when there is a spinner registered.
+
+                var state = new State(Interlocked.Add(ref lockObj._state, Neg(SpinnerCountIncrement)));
+                Debug.Assert(new State(state._state + SpinnerCountIncrement).HasAnySpinners);
+
+                while (true)
+                {
+                    Debug.Assert(state.HasAnyWaiters || !state.ShouldNotPreemptWaiters);
+                    if (state.IsLocked)
+                    {
+                        return TryLockResult.Wait;
+                    }
+
+                    State newState = state;
+                    newState.SetIsLocked();
+
+                    State stateBeforeUpdate = CompareExchange(lockObj, newState, state);
+                    if (stateBeforeUpdate == state)
+                    {
+                        return TryLockResult.Locked;
+                    }
+
+                    state = stateBeforeUpdate;
```

Comment:
Another case of state change collision.

Reply:
Responded here: https://github.com/dotnet/runtime/pull/87672#discussion_r1237335803